src/components/CreateVmWizard/helpers.js

```diff
@@ -1,22 +1,67 @@
-import { createTemplateList } from '_/components/utils'
+import { createIsoList, createTemplateList } from '_/components/utils'
 import timezones from '_/components/utils/timezones.json'
 import { isWindows } from '_/helpers'
 
-const handleClusterIdChange = (clusterId, blankTemplateId, { templates, clusters, defaultValues, operatingSystems, defaultGeneralTimezone, defaultWindowsTimezone }) => {
+const handleClusterIdChange = (clusterId, { blankTemplateId, defaultValues, clusters, templates, operatingSystems, storageDomains, defaultGeneralTimezone, defaultWindowsTimezone }) => {
   let changes = {}
   const templateList = clusterId ? createTemplateList(templates, clusterId) : []
   changes.dataCenterId = clusterId ? clusters.getIn([clusterId, 'dataCenterId']) : undefined
   changes.clusterId = clusterId
-  changes.provisionSource =
-    (templateList.length === 1 && templateList[0].id === blankTemplateId)
-      ? 'iso'
-      : 'template'
-  changes.templateId = templateList.length === 1 ? blankTemplateId : undefined
-  changes = { ...changes, ...handleProvisionSourceChange(changes.provisionSource, defaultValues, { defaultGeneralTimezone, defaultWindowsTimezone, templates, operatingSystems }) }
+
+  if (clusterId === undefined) {
+    changes = { ...changes, ...handleProvisionSourceChange('template', { defaultValues, defaultGeneralTimezone, defaultWindowsTimezone, templates, operatingSystems }) }
+    return changes
+  }
+
+  if (templateList.length > 1) {
+    // cluster includes more than one non blank template
+    changes = {
+      ...changes,
+      ...handleProvisionSourceChange('template', { defaultValues, defaultGeneralTimezone, defaultWindowsTimezone, templates, operatingSystems }),
+    }
+    if (templateList.length === 2) {
+      // cluster includes only one non blank template
+      const selectedTemplate = templateList.find(t => t.id !== blankTemplateId)
+      changes = {
+        ...changes,
+        ...(selectedTemplate ? handleTemplateIdChange(selectedTemplate.id, defaultValues.optimizedFor, { templates, operatingSystems, defaultGeneralTimezone, defaultWindowsTimezone }) : {}),
+      }
+    }
+    return changes
+  }
+  if (templateList.length === 1 && templateList[0].id === blankTemplateId) {
+    // cluster includes only one template (blank) always has blank template
+    const isoList = createIsoList(storageDomains, changes.dataCenterId)
+
+    if (isoList.length > 0) {
+      // cluster includes at least one iso cd
+      changes = {
+        ...changes,
+        ...handleProvisionSourceChange('iso', { defaultValues, defaultGeneralTimezone, defaultWindowsTimezone, templates, operatingSystems }),
+      }
+      if (isoList.length === 1) {
+        // cluster includes only one iso cd
+        changes.isoImage = isoList[0].file.id
+      }
+    } else {
+      // cluster not includes any non blank template or iso cd
```

Comment:
I would change this comment to "Cluster does not include any non blank template".

Regarding the "or iso cd" part of this comment, are you sure about that? Could you please explain? I don't see checking `isoList` here, only in the `if (templateList.length === 1 && templateList[0].id === blankTemplateId)` block.

Reply:
@hstastna Sure, `isoList.length` range is from 0 to n (n = positive number greater than 0 | 1,2,etc...)
In line 36 we are checking if `isoList.length` value is greater than zero:
https://github.com/oVirt/ovirt-web-ui/blob/8c4c7e3fdbb6cd84890e7e2a7472fe070b28279f/src/components/CreateVmWizard/helpers.js#L36
If the result of that condition is `false` so we're jumping into the else block and it means that `isoList.length === 0`, in other words there is no iso for that cluster